### PR TITLE
Add various type-related fixes

### LINKED ISSUE #565: Pointers from array decay can't be directly added to
```
void foo(char a[128]) {char *c = a + 127;}
```
Example benchmark: `Regression/C/gcc-c-torture/execute/20020503-1.c`

### LINKED ISSUE #567: Field access on typedef doesn't work
- `Regression/C/gcc-c-torture/execute/ieee/pr72824-2.c`
- `Regression/C/gcc-c-torture/execute/921113-1.c` (this might be an old C quirk?)
- `Regression/C/gcc-c-torture/execute/pr53645.c`
- `Regression/C/gcc-c-torture/execute/pr53645-2.c`
- `Regression/C/gcc-c-torture/execute/pr70903.c`
- `Regression/C/gcc-c-torture/execute/pr71626-1.c`
- `Regression/C/gcc-c-torture/execute/pr71626-2.c`
- `Regression/C/gcc-c-torture/execute/pr85169.c`

### LINKED ISSUE #569: Subtraction on label pointers (?) doesn't work
I don't believe anyone should ever support this, but it's still a failing test
- `Regression/C/gcc-c-torture/execute/pr70460.c` (Open at your own risk. I warned you.)

### PR DESCRIPTION
**Pull request by Jezurko**
Fix #565
Fix #567 
Fix #569

### CODE REVIEW

File: include/vast/Dialect/HighLevel/HighLevelTypes.td [hunk 1]
```diff
@@ -165,6 +165,20 @@ def ParenType : HighLevelType< "Paren", [ElementTypeInterface, MemRefElementType
...
 }
 
+class ElaboratedType< TypeConstraint value > : Type<
+  And< [
+    ElaboratedType.predicate,
+    SubstLeaves< "$_self", "$_self.cast< ElaboratedType >().getElementType()", value.predicate >
+  ] >
+>;
+
+class MaybeElaboratedType< TypeConstraint value > : Type<
```
Comment (xlauko) [inline include/vast/Dialect/HighLevel/HighLevelTypes.td:175]:
```suggestion
class TypeOrElaboratedTypeOf< TypeConstraint value > : Type<
```

File: include/vast/Dialect/HighLevel/HighLevelTypes.td [hunk 2]
```diff
@@ -207,10 +221,12 @@ class LValueOrType<TypeConstraint value> : Type<
   Or< [
     And< [
       LValue.predicate,
-      SubstLeaves< "$_self", "$_self.cast< LValueType >().getElementType()", value.predicate >
+      SubstLeaves< "$_self", "$_self.cast< LValueType >().getElementType()",
```
Comment (xlauko) [inline include/vast/Dialect/HighLevel/HighLevelTypes.td:224]: Same here, check if cast does not kill the program.

File: lib/vast/CodeGen/DefaultTypeVisitor.cpp [hunk 3]
```diff
@@ -272,6 +281,20 @@ namespace vast::cg {
...
 
+    mlir_type default_type_visitor::VisitVectorType(const clang::VectorType *ty) {
+        return VisitVectorType(ty, clang_qualifiers());
+    }
+
+    mlir_type default_type_visitor::VisitVectorType(const clang::VectorType *ty, clang_qualifiers quals) {
+        return with_cvr_qualifiers(
+            compose_type< hl::VectorType >()
+                // bind also uninitialized size
+                .bind_always(get_size_attr(ty, self.mcontext()))
```
Comment (xlauko) [inline lib/vast/CodeGen/DefaultTypeVisitor.cpp:292]: This is always unitialized, or I don't get the comment.

`get_size_attr` always creates initialized `hl::SizeParam`

File: include/vast/Dialect/HighLevel/HighLevelTypes.td [hunk 4]
```diff
@@ -165,6 +165,20 @@ def ParenType : HighLevelType< "Paren", [ElementTypeInterface, MemRefElementType
   let assemblyFormat = "`<` $elementType `>`";
 }
 
+class ElaboratedType< TypeConstraint value > : Type<
+  And< [
+    ElaboratedType.predicate,
+    SubstLeaves< "$_self", "$_self.cast< ElaboratedType >().getElementType()", value.predicate >
```
[the remark this answers is not in the document]
Comment (Jezurko) [reply include/vast/Dialect/HighLevel/HighLevelTypes.td:171]: This should work, as the generated code chains the predicates in `And` via `&&`. From looking at the code it seems to be fine (although it's a 1000 character long condition, so take it with a grain of salt)

File: lib/vast/CodeGen/DefaultTypeVisitor.cpp (hunk 3 above)
Comment (Jezurko) [reply lib/vast/CodeGen/DefaultTypeVisitor.cpp:292]: This is a mistake, this comment should not be here

**Review by xlauko — APPROVED** [review]

**Pull request merged**